In [1]:
import pandas as pd
import numpy as np

df = pd.read_csv('Electric_Vehicle_Population_Data.csv')
print("Dataset loaded successfully!")

df.head()

Dataset loaded successfully!


VIN (1-10)     County       City State  Postal Code  Model Year   Make  \
0  5YJYGDEE1L       King    Seattle    WA      98122.0        2020  TESLA   
1  7SAYGDEE9P  Snohomish    Bothell    WA      98021.0        2023  TESLA   
2  5YJSA1E4XK       King    Seattle    WA      98109.0        2019  TESLA   
3  5YJSA1E27G       King   Issaquah    WA      98027.0        2016  TESLA   
4  5YJYGDEE5M     Kitsap  Suquamish    WA      98392.0        2021  TESLA   

     Model           Electric Vehicle Type  \
0  MODEL Y  Battery Electric Vehicle (BEV)   
1  MODEL Y  Battery Electric Vehicle (BEV)   
2  MODEL S  Battery Electric Vehicle (BEV)   
3  MODEL S  Battery Electric Vehicle (BEV)   
4  MODEL Y  Battery Electric Vehicle (BEV)   

   Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
0            Clean Alternative Fuel Vehicle Eligible             291   
1  Eligibility unknown as battery range has not b...               0   
2            Clean Alternative Fuel Vehicle Eligible             270   
3            Clean Alternative Fuel Vehicle Eligible             210   
4  Eligibility unknown as battery range has not b...               0   

   Base MSRP  Legislative District  DOL Vehicle ID  \
0          0                  37.0       125701579   
1          0                   1.0       244285107   
2          0                  36.0       156773144   
3          0                   5.0       165103011   
4          0                  23.0       205138552   

                Vehicle Location  \
0   POINT (-122.30839 47.610365)   
1  POINT (-122.179458 47.802589)   
2   POINT (-122.34848 47.632405)   
3   POINT (-122.03646 47.534065)   
4   POINT (-122.55717 47.733415)   

                                Electric Utility  2020 Census Tract  
0   CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)       5.303301e+10  
1                         PUGET SOUND ENERGY INC       5.306105e+10  
2   CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)       5.303301e+10  
3  PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)       5.303303e+10  
4                         PUGET SOUND ENERGY INC       5.303594e+10

In [2]:
print("\n--- DataFrame Info ---")
df.info()


--- DataFrame Info ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 177866 entries, 0 to 177865
Data columns (total 17 columns):
 #   Column                                             Non-Null Count   Dtype  
---  ------                                             --------------   -----  
 0   VIN (1-10)                                         177866 non-null  object 
 1   County                                             177861 non-null  object 
 2   City                                               177861 non-null  object 
 3   State                                              177866 non-null  object 
 4   Postal Code                                        177861 non-null  float64
 5   Model Year                                         177866 non-null  int64  
 6   Make                                               177866 non-null  object 
 7   Model                                              177866 non-null  object 
 8   Electric Vehicle Type                             

In [3]:
print("\nDescriptive Statistics")
print(df.describe())


Descriptive Statistics
         Postal Code     Model Year  Electric Range      Base MSRP  \
count  177861.000000  177866.000000   177866.000000  177866.000000   
mean    98172.453506    2020.515512       58.842162    1073.109363   
std      2442.450668       2.989384       91.981298    8358.624956   
min      1545.000000    1997.000000        0.000000       0.000000   
25%     98052.000000    2019.000000        0.000000       0.000000   
50%     98122.000000    2022.000000        0.000000       0.000000   
75%     98370.000000    2023.000000       75.000000       0.000000   
max     99577.000000    2024.000000      337.000000  845000.000000   

       Legislative District  DOL Vehicle ID  2020 Census Tract  
count         177477.000000    1.778660e+05       1.778610e+05  
mean              29.127481    2.202313e+08       5.297672e+10  
std               14.892169    7.584987e+07       1.578047e+09  
min                1.000000    4.385000e+03       1.001020e+09  
25%               18

In [4]:
print("\n--- Unique Values in Key Columns ---")
for col in ['State', 'Electric Vehicle Type', 'Make', 'Model Year']:
    print(f"{col}: {df[col].nunique()} unique values")
    if df[col].nunique() < 20:
        print(df[col].unique())



--- Unique Values in Key Columns ---
State: 46 unique values
Electric Vehicle Type: 2 unique values
['Battery Electric Vehicle (BEV)' 'Plug-in Hybrid Electric Vehicle (PHEV)']
Make: 40 unique values
Model Year: 22 unique values


**DATASET OBSERVATIONS**

1 - Dataset Size: 177,866 records and 17 columns. 

2 - Missing Values (Non-Null Counts):

-  County, City, Postal Code, Electric Utility, 2020 Census Tract: 5 missing values each. These are very few.
- Legislative District: 389 missing values. A significant number.
- Vehicle Location: 9 missing values. Also very few.

3 - Data Types:
- Postal Code, Legislative District, 2020 Census Tract are currently float64 but, are more likely identifiers or categorical. The float type is usually due to the presence of NaN values, which Pandas represents as floats.
- DOL Vehicle ID is int64, should change to object as a unique identifier.

4 - Descriptive Statistics:
- Model Year: Range 1997-2024, mean 2020.5. A rich historical dataset.
- Electric Range: Mean: 58.8 miles. 25th, 50th (median), and 75th percentiles are all 0. This is a major data quality issue. This strongly suggests missing data, or perhaps miscoded data where 0 means "not applicable".
- Base MSRP: Mean: $1,073. 25th, 50th (median), and 75th percentiles are all $0. This indicates widespread missing or placeholder data rather than actual $0 prices.

5 - Unique Values in Categorical Columns:
- State: 46 unique states
- Electric Vehicle Type: BEV, PHEV
- Make: 40 unique makes – good for market share analysis.
- Model Year: 22 unique years – good for temporal analysis.

--------------------------------------------------------------------------------------
**Cleaning next steps:**

- Fix electric range data quality issues
- Address missing MSRP data
- Review location data for the 5 missing records
- Imputation strategies for legislative district missing values




In [5]:
# Convert DOL Vehicle ID to string
df['DOL Vehicle ID'] = df['DOL Vehicle ID'].astype(str)

In [ ]:
# Drop na for a very few missing values

cols_to_drop_na = ['County', 'City', 'Postal Code', 'Vehicle Location', 'Electric Utility', '2020 Census Tract']

initial_rows = df.shape[0]
df.dropna(subset=cols_to_drop_na, inplace=True)

In [7]:
# Treat Legislative District missing values.

# Impute with a placeholder (-1)
df['Legislative District'] = df['Legislative District'].fillna(-1).astype(str)


In [ ]:
# Convert Postal Code and 2020 Census Tract to string/object type
df['Postal Code'] = df['Postal Code'].astype(str)
df['2020 Census Tract'] = df['2020 Census Tract'].astype(str)



Data types after converting Postal Code and 2020 Census Tract:
Postal Code             object
2020 Census Tract       object
Legislative District    object
dtype: object


In [9]:
# Address 0 values in Base MSRP
initial_zero_msrp_count = df[df['Base MSRP'] == 0].shape[0]
df['Base MSRP'].replace(0, pd.NA, inplace=True)
df['Base MSRP'] = df['Base MSRP'].astype('Float64')

print(f"\nConverted {initial_zero_msrp_count} '0' values in 'Base MSRP' to NaN (pd.NA).")



Converted 174514 '0' values in 'Base MSRP' to NaN (pd.NA).


C:\Users\luket\AppData\Local\Temp\ipykernel_27964\1234442314.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Base MSRP'].replace(0, pd.NA, inplace=True)


**With 174514 missing values in 'Base MSRP', this column is heavily missing. The decision is to entirely drop this column**

In [10]:
# Drop "Base MSRP" column
df.drop('Base MSRP', axis=1, inplace=True)
print("Base MSRP column dropped.")

Base MSRP column dropped.


In [11]:
# Address 0 values in Electric Range
print("\nCounts of 0 Electric Range by Electric Vehicle Type:")
print(df[df['Electric Range'] == 0]['Electric Vehicle Type'].value_counts())

initial_bev_zero_range_count = df[(df['Electric Range'] == 0) & (df['Electric Vehicle Type'] == 'Battery Electric Vehicle (BEV)')].shape[0]
df.loc[(df['Electric Range'] == 0) & (df['Electric Vehicle Type'] == 'Battery Electric Vehicle (BEV)'), 'Electric Range'] = pd.NA
print(f"Converted {initial_bev_zero_range_count} '0' Electric Range for BEVs to NaN (pd.NA).")


Counts of 0 Electric Range by Electric Vehicle Type:
Electric Vehicle Type
Battery Electric Vehicle (BEV)    91949
Name: count, dtype: int64
Converted 91949 '0' Electric Range for BEVs to NaN (pd.NA).


**91,949 BEV records have NaN for Electric Range. This is too many to drop. The decision here is Median imputation based on 'Make' and 'Model Year'**

In [12]:
print("\n--- Imputing Missing Electric Range for BEVs ---")
df['Imputed Electric Range'] = df.groupby(['Make', 'Model Year'])['Electric Range'].transform('median')

initial_nan_bev_range_count = df.loc[(df['Electric Range'].isna()) & (df['Electric Vehicle Type'] == 'Battery Electric Vehicle (BEV)')].shape[0]

df.loc[(df['Electric Range'].isna()) & (df['Electric Vehicle Type'] == 'Battery Electric Vehicle (BEV)'), 'Electric Range'] = \
    df.loc[(df['Electric Range'].isna()) & (df['Electric Vehicle Type'] == 'Battery Electric Vehicle (BEV)'), 'Imputed Electric Range']

overall_bev_median_range = df[df['Electric Vehicle Type'] == 'Battery Electric Vehicle (BEV)']['Electric Range'].median()

df['Electric Range'].fillna(overall_bev_median_range, inplace=True)

df.drop('Imputed Electric Range', axis=1, inplace=True)



--- Imputing Missing Electric Range for BEVs ---


C:\Users\luket\AppData\Local\Temp\ipykernel_27964\2294630861.py:11: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Electric Range'].fillna(overall_bev_median_range, inplace=True)


In [13]:
df.isnull().sum()

VIN (1-10)                                           0
County                                               0
City                                                 0
State                                                0
Postal Code                                          0
Model Year                                           0
Make                                                 0
Model                                                0
Electric Vehicle Type                                0
Clean Alternative Fuel Vehicle (CAFV) Eligibility    0
Electric Range                                       0
Legislative District                                 0
DOL Vehicle ID                                       0
Vehicle Location                                     0
Electric Utility                                     0
2020 Census Tract                                    0
dtype: int64

In [14]:
print(df.describe())

          Model Year  Electric Range
count  177857.000000   177857.000000
mean     2020.515684      124.630490
std         2.989208       77.021998
min      1997.000000        6.000000
25%      2019.000000       38.000000
50%      2022.000000      151.000000
75%      2023.000000      151.000000
max      2024.000000      337.000000


In [15]:
output_file_path = 'ev_population_data_cleaned.csv'
df.to_csv(output_file_path, index=False)
print(f"\nCleaned data saved to: {output_file_path}")


Cleaned data saved to: ev_population_data_cleaned.csv
